llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -9119,6 +9121,23 @@ bool llvm::matchSimpleRecurrence(const BinaryOperator *I, PHINode *&P,
   return P && matchSimpleRecurrence(P, BO, Start, Step) && BO == I;
 }
 
+bool llvm::matchSimpleBinaryIntrinsicRecurrence(const IntrinsicInst *I,
+                                                PHINode *&P, Value *&Init,
+                                                Value *&Invariant) {
+  IntrinsicInst *II = nullptr;
+  P = dyn_cast<PHINode>(I->getArgOperand(0));
+  if (!P)
+    P = dyn_cast<PHINode>(I->getArgOperand(1));
+
+  std::function<bool(const IntrinsicInst *)> IsBinaryOrMinMaxFn =
```

Comment:
Can we run the predicate after matching the recurrence?

Reply:
We could run the predicate after matching the recurrence, but that would imply computing all the recurrence needlessly, if the intrinsic turned out not to be a candidate one (as we're passing a general IntrinsicInst), in addition to checking intrinsic operands size (as not to go out of bound when retrieving the Operation ops). Switched this to using a lightweight llvm::function_ref instead, I think it should be quite better.